### PR TITLE
Fix invalid handling of pids.limit=0 from runtime spec

### PR DESCRIPTION
**Pull request by kolyshkin**
_This is an alternative to #4015 (smaller set of code changes, no libct API breakage)._

It has been pointed out in #4014 that runc incorrectly ignores pids.limit=0 set
in the runtime spec. This happens because runtime-spec says "default is
unlimited" and also allows for Pids to not be set at all, thus
distinguishing unset (Resources.Pids == nil) from unlimited
(Resources.Pids.Limit <= 0).

Internally, runc also distinguishes unset from unlimited, but since we
do not use a pointer, we treat 0 as unset and -1 as unlimited.

Add a conversion code to libcontainer/specconv.

Add a test case to check that starting a container with pids.limit=0
results in no pids limit (the test fails before the fix when systemd
cgroup manager is used, as systemd apparently uses parent's TasksMax;
see #4022 for CI runs).

NOTE that runc update still treats 0 as "unset".

Finally, fix/update the documentation here and there.

Fixes: #4014.

### CODE REVIEW

**Comment by kolyshkin** [comment]
@haircommander PTAL

File: update.go
Comment (lifubang) [inline update.go:129]: Maybe some users will be confused with `-r` and `runtime-spec`?
```
runc update -r - containerid
{
  "pids": { "limit": 0 }
}
```

File: libcontainer/specconv/spec_linux.go
Comment (lifubang) [inline libcontainer/specconv/spec_linux.go:777]: I don't know whether upstream projects should write such codes or not when they update the container's pids limit?
```
if r.Pids.Limit > 0 {
    c.Resources.PidsLimit = r.Pids.Limit
} else {
    c.Resources.PidsLimit = -1
}
```
@AkihiroSuda Do you know what's the situation in `nerdctl`.

Comment (lifubang) [reply libcontainer/specconv/spec_linux.go:777]: Maybe in kubelet they should know this information:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cgroup_manager_linux.go#L355-L381

Comment (lifubang) [reply libcontainer/specconv/spec_linux.go:777]: So I still think we should fix it in `pids.go`, because others always use `Manager.Set` directly without use `specconv`.
But we have to solve 2 problems if we choose solutions like #4015 :
1. Don't always write `pids.max` if `PidsLimit == 0`;
2. Don't introduce break changes of libcontainer API.

Comment (kolyshkin) [reply libcontainer/specconv/spec_linux.go:777]: I took a look at git history, and it seems that libcontainer _always_ treated `0` as unset for pids limit, since commit db3159c9d9d3b65d2d27ff33f722eb9797f20e52 from January 2016 that added the initial pids limit support (alas, the very same commit incorrectly documents the field (https://github.com/opencontainers/runc/commit/db3159c9d9d3b65d2d27ff33f722eb9797f20e52#diff-b71b6973c045d22e41e802cf65cade82c573a844190bc0d06a6ade9f21cb2c5c), and this PR fixes that).

Thus, I am pretty sure libcontainer users are aware of what 0 means for pids limit in libcontainer; yet I am going to look into kubernetes and nerdctl to confirm.

Now, we have two problems to solve:
1. runc violates the runtime spec by treating pids.limit==0 as unset rather than unlimited (#4014);
2. libcontainer's handling of pids limit value differs from runtime spec.

I think that 2 is much less of a problem than 1, and can be solved by properly documenting the PidsLimit field (this is what this PR does).

Comment (haircommander) [reply libcontainer/specconv/spec_linux.go:777]: FWIW: case 1 is the only case I care about fixing

**Review by haircommander — APPROVED** [review]
LGTM, thanks!

File: update.go
Comment (kolyshkin) [reply update.go:129]: 1. This is the way it was always working.
2. It seems that the behavior of `runc update` is not in spec.

**Comment by lifubang** [comment]
If we say that we just only need to fix it in ‘runc create’ and ‘runc run’, for the uses using ‘runc update’ and ‘libcontainer’, they should fix this issue in their own projects according to runc’s doc. This one LGTM.

**Comment by kolyshkin** [comment]
> for the uses using ‘runc update’ and ‘libcontainer’,

Yes, both `runc update` and `libcontainer` API is not actually covered by runtime-spec (which is sort of a problem, but out of scope for this).

**Review by lifubang — APPROVED** [review]
LGTM, to fix this issue completely, it needs to open PR by their own contributors in projects using ‘runc update’ or ‘runc libcontainer API’, for example, Kubernetes, containerd, and cri-o etc.

**Comment by lifubang** [comment]
@opencontainers/runc-maintainers PTAL
It's really a bug for runc violates the runtime spec by treating pids.limit==0 as unset rather than unlimited. (https://github.com/opencontainers/runc/pull/4023#discussion_r1337749605)

After this one backported to 1.1, we can release `1.1.10`.

**Comment by lifubang** [comment]
@AkihiroSuda PTAL
I have checked docker cli and nerdctl, they tell users to use -1 for unlimited pids limit.

**Review by AkihiroSuda — APPROVED** [review]

**Comment by thaJeztah** [comment]
I'd yet have to have a close look, but would like @cyphar to have a look; I think there's still some ambiguity in the description of the PR that updated the specs https://github.com/opencontainers/runtime-spec/pull/234/l   ("there is no limit: meaning: don't (set) a limit (and use the active max)" vs "explicitly set it to max/unlimited")

The mention of "Go's unset is zero value" in that description may indicate that `0 == unset`  and thus the former may have been the intent?

And it wouldn't be the first time that specs were update to reflect the implementation, and now (due to ambiguity on the wording), the implementation being update to reflect the spec (going full circle)

**Comment by kolyshkin** [comment]
Rebased

**Comment by cyphar** [comment]
Reading through my (8-year-old :cold_sweat:) comments, I think that there were two issues at play here:

1. A value of `0` being identical to `1` is a quirk of the way pids limits work (a process needs to be in the cgroup to fork into the cgroup, so a limit of `0` and `1` are identical in terms of behaviour -- though with `CLONE_INTO_CGROUP` this is no longer true).
2. When I wrote that PR, I had sent a patch to make `0` an invalid value of a pids cgroup (due to point 1). It was never merged, but I think the logic was that having two values that mean the same thing was not ideal (and the cgroup maintainer had made a similar comment before merging the cgroup).
3. At the time, we hadn't fully settled on the whole pointers-in-Go-JSON thing, and so I just used the zero value of `int64` as a temporary stopgap until we did a tree-wide fixup. It seems we never did a proper tree-wide fixup (or we skipped `PidsLimit`).

Looking at it now, I think the _ideal_ option would be to change the spec so that `PidsLimit` is an `*int64` and then we can have the behaviour of most other cgroups (unset/`nil` means to not touch the limit, any value means "use that value" with `-1` meaning unlimited). However, given the disparate behaviour of runtimes when dealing with `0`, the spec should have a note saying that (prior to whatever release has the new change) the behaviour was not always consistent...

**Comment by kolyshkin** [comment]
Removing the `backport/todo/1.1` as we're not sure how to move forward with this (maybe we'll change the spec instead, since runc never accepted `0` as `unlimited`.

**Comment by lifubang** [comment]
> maybe we'll change the spec instead, since runc never accepted `0` as `unlimited`.

👍

**Comment by haircommander** [comment]
in the meantime, CRI-O can just set -1 where it originally set 0

**Comment by AkihiroSuda** [comment]
What's the next step?

**Comment by kolyshkin** [comment]
Let's close it and change the spec instead,

**Pull request closed without merging**
